include/execpools/thread_pool_base.hpp

```diff
@@ -0,0 +1,485 @@
+          *sch.pool_, static_cast<S&&>(sndr), shape, static_cast<Fun&&>(fun)};
+      }
+
+      [[nodiscard]]
+      constexpr auto
+        forward_progress_guarantee() const noexcept -> stdexec::forward_progress_guarantee {
+        return pool_->forward_progress_guarantee();
+      }
+
+      friend thread_pool_base;
+
+      explicit scheduler(DerivedPoolType& pool) noexcept
+        : pool_(&pool) {
+      }
+
+      DerivedPoolType* pool_;
+
+     public:
+      using __t = scheduler;
+      using __id = scheduler;
+      auto operator==(const scheduler&) const -> bool = default;
+
+      constexpr auto query(stdexec::get_forward_progress_guarantee_t) const noexcept
+        -> stdexec::forward_progress_guarantee {
+        return forward_progress_guarantee();
+      }
+
+      auto schedule() const noexcept -> sender {
+        return sender{*pool_};
+      }
+    };
+
+    [[nodiscard]]
+    auto get_scheduler() noexcept -> scheduler {
+      return scheduler{static_cast<DerivedPoolType&>(*this)};
+    }
+
+    /* Is this even needed? Looks like no.
+     * void request_stop() noexcept {
+        // Should this do anything? TBB supports its own flavor of cancelation at a tbb::task_group level
+        //
+    https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/scheduling_controls/task_group_context_cls.html
+        // but not at the tbb::task_arena level.
+        // https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.html
+    }*/
+
+    [[nodiscard]]
+    auto available_parallelism() const -> std::uint32_t {
+      return static_cast<DerivedPoolType&>(*this).available_parallelism();
+    }
+
+   private:
+    void enqueue(task_base* task, std::uint32_t tid = 0) noexcept {
+      static_cast<DerivedPoolType&>(*this).enqueue(task, tid);
+    }
+
+    void bulk_enqueue(task_base* task, std::uint32_t n_threads) noexcept {
+      for (std::uint32_t tid = 0; tid < n_threads; ++tid) {
+        this->enqueue(task, tid);
+      }
+    }
+  };
+
+  template <class PoolType, class ReceiverId>
+  struct operation<PoolType, ReceiverId>::__t : task_base {
+    using __id = operation;
+    friend class thread_pool_base<PoolType>;
+
+    PoolType& pool_;
+    Receiver rcvr_;
+
+    explicit __t(PoolType& pool, Receiver rcvr)
+      : pool_(pool)
+      , rcvr_(std::move(rcvr)) {
+      this->__execute =
+        [](task_base* t, std::uint32_t /* tid What is this needed for? */) noexcept {
+          auto& op = *static_cast<__t*>(t);
+          auto stoken = stdexec::get_stop_token(stdexec::get_env(op.rcvr_));
+          if (stoken.stop_requested()) {
+            stdexec::set_stopped(std::move(op.rcvr_));
+          } else {
+            stdexec::set_value(std::move(op.rcvr_));
+          }
+        };
+    }
+
+    void enqueue() noexcept {
+      pool_.enqueue(this);
+    }
+
+    void start() & noexcept {
+      enqueue();
+    }
+  };
+} // namespace execpools
```

Comment:
A good test/useage-example of this would be like an `AnyThreadPool` where tasks are submitted as `std::function<void()>`s.

Reply:
I agree, more example makes it more usable! I just would like to questioning back, to clarify, if you don't mind.

The idea is to have an `AnyThreadPool` that based on the `thread_pool_base` but doesn't have a `get_scheduler()` public function to schedule tasks on it, but would have an `enqueue(std::function<void()> callable)` function how one can submit task to the pool. Is that correct?

Reply:
Something like that.  When I wrote this originally I was coming at it with the notion that a thread-based execution context must be modeled by something g you throw `std::function<void()>`s at and that it promises to execute eventually. (Or it could be `void(*f)(void*)`s and corresponding `void* x` as a pair and it promises to have a thread call `f(x)`. That’s weaker and I think still fits.)